src/rust/lib/graph-editor/src/component/node/port/output.rs

```diff
@@ -0,0 +1,365 @@
+///  * when one of the output ports is hovered, after a set time, all ports are show and the hovered
+///    port is highlighted.
+///  * when a different port is hovered, it is highlighted immediately.
+///  * when none of the ports is hovered all of the `OutputPorts` disappear. Note: there is a very
+///    small delay for disappearing to allow for smooth switching between ports.
+///
+#[derive(Debug,Clone,CloneRef)]
+pub struct OutputPorts {
+    /// The FRP api of the `OutPutPorts`.
+    pub frp     : Frp,
+        network : frp::Network,
+        data    : Rc<OutputPortsData>,
+}
+
+impl OutputPorts {
+    /// Constructor.
+    pub fn new(scene:&Scene, number_of_ports:u32) -> Self {
+        let network = default();
+        let frp     = Frp::new(&network);
+        let data    = OutputPortsData::new(scene.clone_ref(), number_of_ports);
+        let data    = Rc::new(data);
+        OutputPorts {data,network,frp}.init()
+    }
+
+    fn init(mut self) -> Self {
+        self.init_frp();
+        self
+    }
+
+    fn init_frp(&mut self) {
+        let network = &self.network;
+        let frp     = &self.frp;
+        let data    = &self.data;
+
+        // Used to set and detect the end of the tweens. The actual value is irrelevant, only the
+        // duration of the tween matters and that this value is reached after that time.
+        const TWEEN_END_VALUE:f32 = 1.0;
+
+        // Timer used to measure whether the hover has been long enough to show the ports.
+        let delay_show = Tween::new(&network);
+        delay_show.set_duration(SHOW_DELAY_DURATION);
+
+        // Timer used to measure whether the mouse has been gone long enough to hide all ports.
+        let delay_hide = Tween::new(&network);
+        delay_hide.set_duration(HIDE_DELAY_DURATION);
+
+        frp::extend! { network
+
+            // === Size Change Handling == ///
+
+            eval frp.set_size ((size) data.set_size(size.into()));
+
+
+            // === Hover Event Handling == ///
+
+            port_mouse_over      <- source::<PortId>();
+            port_mouse_out       <- source::<PortId>();
+
+            delay_show_finished    <- delay_show.value.map(|t| *t>=TWEEN_END_VALUE );
+            delay_hide_finished    <- delay_hide.value.map(|t| *t>=TWEEN_END_VALUE );
+            on_delay_show_finished <- delay_show_finished.gate(&delay_show_finished).constant(());
+            on_delay_hide_finished <- delay_hide_finished.gate(&delay_hide_finished).constant(());
+
+            visible                <- delay_show_finished.map(|v| *v);
+
+            mouse_over_while_inactive  <- port_mouse_over.gate_not(&visible).constant(());
+            mouse_over_while_active    <- port_mouse_over.gate(&visible).constant(());
+
+            eval mouse_over_while_inactive ([delay_show,delay_hide](_){
+                delay_hide.stop();
+                delay_show.rewind();
+                delay_show.set_end_value(TWEEN_END_VALUE);
```

Comment:
why this is needed?

Reply:
We need to cancel racing timer to avoid both timers ending.